lib/codegen/meta/src/cdsl/typevar.rs

```diff
@@ -0,0 +1,345 @@
+            && self.specials.is_subset(&other.specials)
+    }
+
+    pub fn as_bool(&self) -> TypeSet {
+        let mut new = self.clone();
+        new.ints.clear();
+        new.floats.clear();
+        new.bitvecs.clear();
+
+        if self
+            .lanes
+            .difference(&[1usize].to_vec().into_iter().collect::<IndexSet<usize>>())
+            .count()
+            > 0
+        {
+            new.bools = self
+                .ints
+                .union(&self.floats)
+                .map(|i| *i)
+                .collect::<IndexSet<usize>>()
+                .union(&self.bools)
+                .map(|i| *i)
+                .collect::<IndexSet<usize>>();
+        }
+
+        if self.lanes.contains(&1) {
+            new.bools.insert(1);
+        }
+
+        new
+    }
+}
+
+#[derive(Default)]
+pub struct TypeSetBuilder {
+    lanes: Option<Interval>,
+    ints: Option<Interval>,
+    floats: Option<Interval>,
+    bools: Option<Interval>,
+    bitvecs: Option<Interval>,
+    specials: Option<SpecialSpec>,
+}
+
+impl TypeSetBuilder {
+    pub fn lanes(mut self, interval: Interval) -> Self {
+        self.lanes = Some(interval);
+        self
+    }
+
+    pub fn ints(mut self, interval: Interval) -> Self {
+        self.ints = Some(interval);
+        self
+    }
+
+    pub fn floats(mut self, interval: Interval) -> Self {
+        self.floats = Some(interval);
+        self
+    }
+
+    pub fn bools(mut self, interval: Interval) -> Self {
+        self.bools = Some(interval);
+        self
+    }
+
+    pub fn bitvecs(mut self, interval: Interval) -> Self {
+        self.bitvecs = Some(interval);
+        self
+    }
+
+    pub fn specials(mut self, specials: SpecialSpec) -> Self {
+        self.specials = Some(specials);
+        self
+    }
+
+    pub fn finish(self) -> TypeSet {
+        TypeSet {
+            lanes: Interval::to_iter(Interval::decode(self.lanes, 1..MAX_LANES, Some(1))).collect(),
+            ints: Interval::to_iter(Interval::decode(self.ints, 8..MAX_BITS, None)).collect(),
+            floats: Interval::to_iter(Interval::decode(self.floats, 32..64, None)).collect(),
+            bools: Interval::to_iter(Interval::decode(self.bools, 1..MAX_BITS, None))
+                .filter(|&bits| {
+                    bits == 1 || (bits >= 8 && bits <= MAX_BITS && bits.is_power_of_two())
+                }).collect(),
+            bitvecs: Interval::to_iter(Interval::decode(self.bitvecs, 1..MAX_BITVEC, None))
+                .collect(),
+            specials: SpecialSpec::to_set(self.specials),
+        }
+    }
+}
+
+pub struct TypeVar {
+    name: String,
+    doc: &'static str,
+    type_set: TypeSet,
+    base: Option<Box<TypeVar>>,
+    derived_func: Option<&'static str>,
+}
+
+impl TypeVar {
+    pub fn build(name: impl Into<String>) -> TypeVarBuilder {
+        TypeVarBuilder {
+            name: name.into(),
+            doc: "",
+            type_set_builder: TypeSet::build(),
```

Comment:
We should decide on `TypeSet::build` vs `TypeSetBuilder::new`, or `Instruction`/`InstructionBuilder`, because they're currently different from each other and it'd be good to pick one way or the other :-).

I know it's more verbose, but I slightly prefer `TypeSetBuilder::new`. It's my vague impression that this style is more common in Rust code, though I'm open to learning otherwise.